From 72b14b600eee6b3c5dd05e60e5ba9dbfbe03795b Mon Sep 17 00:00:00 2001
From: Fabien Dessenne <fabien.dessenne@foss.st.com>
Date: Thu, 8 Dec 2022 17:32:33 +0100
Subject: [PATCH] media: v4l: add support of STM32 ISP location/average/bin
 controls

Add the following ISP statistics controls:
- location: control where the stats are captured
- average filter: control the average stat filter strategy
- bin component: control the component source of the bin stat

Signed-off-by: Fabien Dessenne <fabien.dessenne@foss.st.com>
Change-Id: I18525f1e00f31a612ab340b71a71ed6c97bb7371
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/292770
Tested-by: Eric FOURMONT <eric.fourmont-ext@st.com>
Reviewed-by: Eric FOURMONT <eric.fourmont-ext@st.com>
Domain-Review: Eric FOURMONT <eric.fourmont-ext@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/media/v4l2-core/v4l2-ctrls-defs.c |  3 +++
 include/uapi/linux/v4l2-controls.h        | 22 ++++++++++++++++++++++
 2 files changed, 25 insertions(+)

--- a/drivers/media/v4l2-core/v4l2-ctrls-defs.c
+++ b/drivers/media/v4l2-core/v4l2-ctrls-defs.c
@@ -1164,6 +1164,9 @@ const char *v4l2_ctrl_get_name(u32 id)
 	case V4L2_CID_DEINTERLACING_MODE:	return "Deinterlacing Mode";
 	case V4L2_CID_DIGITAL_GAIN:		return "Digital Gain";
 	case V4L2_CID_ISP_STAT_REGION:		return "ISP Stat Region";
+	case V4L2_CID_ISP_STAT_LOCATION:	return "ISP Stat Location";
+	case V4L2_CID_ISP_STAT_AVG_FILTER:	return "ISP Stat Average Filter";
+	case V4L2_CID_ISP_STAT_BIN_COMP:	return "ISP Stat Bin Component";
 
 	/* DV controls */
 	/* Keep the order of the 'case's the same as in v4l2-controls.h! */
--- a/include/uapi/linux/v4l2-controls.h
+++ b/include/uapi/linux/v4l2-controls.h
@@ -1230,6 +1230,28 @@ struct v4l2_ctrl_isp_stat_region {
 	__u32 height[25];
 };
 
+#define V4L2_CID_ISP_STAT_LOCATION		(V4L2_CID_IMAGE_PROC_CLASS_BASE + 11)
+enum v4l2_isp_stat_location {
+	V4L2_STAT_LOCATION_BEFORE_PROC	= 0,
+	V4L2_STAT_LOCATION_AFTER_DEMO	= 1,
+};
+
+#define V4L2_CID_ISP_STAT_AVG_FILTER		(V4L2_CID_IMAGE_PROC_CLASS_BASE + 12)
+enum v4l2_isp_stat_avg_filter {
+	V4L2_STAT_AVG_FILTER_NONE	= 0,
+	V4L2_STAT_AVG_FILTER_EXCL16	= 1,
+	V4L2_STAT_AVG_FILTER_EXCL32	= 2,
+	V4L2_STAT_AVG_FILTER_EXCL64	= 3,
+};
+
+#define V4L2_CID_ISP_STAT_BIN_COMP		(V4L2_CID_IMAGE_PROC_CLASS_BASE + 13)
+enum v4l2_isp_stat_bin_comp {
+	V4L2_STAT_BIN_COMP_R		= 0,
+	V4L2_STAT_BIN_COMP_G		= 1,
+	V4L2_STAT_BIN_COMP_B		= 2,
+	V4L2_STAT_BIN_COMP_L		= 3,
+};
+
 /*  DV-class control IDs defined by V4L2 */
 #define V4L2_CID_DV_CLASS_BASE			(V4L2_CTRL_CLASS_DV | 0x900)
 #define V4L2_CID_DV_CLASS			(V4L2_CTRL_CLASS_DV | 1)
